File: leetCode/Python/206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # iteratively(two pointer)
        prev = None
        while head:
            cur = head
            head = head.next
            cur.next = prev
            prev = cur
        return prev 
        
        # recursively
        # if head == None or head.next == None:
        #     return head
        # ans = self.reverseList(head.next)
        # # this is hard to understand
        # head.next.next = head
        # head.next = None
        # return ans

File: leetCode/CSharp/206-reverse-linked-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseList(ListNode head) {
        if (head == null) {
            return null;
        }
        // two pointer
        ListNode result = null;
        while (head != null) {
            ListNode cur = head;
            head = head.next;
            cur.next = result;
            result = cur;
        }
        return result;
    }
}